Add Dell PowerFlex driver (#12304)
canonical/lxd: doc/reference/storage_powerflex.md

(storage-powerflex)=
# Dell PowerFlex - `powerflex`

[Dell PowerFlex](https://www.dell.com/en-us/dt/storage/powerflex.htm) is a software-defined storage solution from [Dell Technologies](https://dell.com). Among other things it offers the consumption of redundant block storage across the network.

LXD offers access to PowerFlex storage clusters by making use of the NVMe/TCP transport protocol.
In addition, PowerFlex offers copy-on-write snapshots, thin provisioning and other features.

To use PowerFlex, make sure you have the required kernel modules installed on your host system.

> On Ubuntu these are `nvme_fabrics` and `nvme_tcp` which come bundled in the `linux-modules-extra-$(uname -r)` package.

Reviewer: ```suggestion
On Ubuntu these are `nvme_fabrics` and `nvme_tcp`, which come bundled in the `linux-modules-extra-$(uname -r)` package.
```

## Terminology

> PowerFlex groups various so called SDS (storage data servers) under logical groups within a protection domain.
> Those SDS are the hosts contributing storage capacity to the PowerFlex cluster.
> A protection domain contains storage pools which represent a set of physical storage devices from different SDS.
> In those storage pools LXD creates its volumes.

Reviewer: ```suggestion
PowerFlex groups various so-called {abbr}`SDS (storage data servers)` under logical groups within a protection domain.
Those SDS are the hosts that contribute storage capacity to the PowerFlex cluster.
A *protection domain* contains storage pools, which represent a set of physical storage devices from different SDS.
LXD creates its volumes in those storage pools.
```

We use italics to introduce new terminology. "SDS" should really be italics as well, but unfortunately we can't combine that with the `{abbr}`, which I think is nicer to have here. :)

> From any volume in PowerFlex a snapshot can be taken which will create an independent copy of the parent volume.

Reviewer: ```suggestion
You can take a snapshot of any volume in PowerFlex, which will create an independent copy of the parent volume.
```
Not sure if "you" is correct here? But if we can avoid passive voice, we should.

Author: "You" is perfectly fine here.

PowerFlex volumes get added as a NVMe drive to the respective LXD host the volume got mapped to.

> For this the LXD host connects to one or multiple NVMe storage data targets (SDT) provided by PowerFlex.

Reviewer: ```suggestion
For this, the LXD host connects to one or multiple NVMe {abbr}`SDT (storage data targets)` provided by PowerFlex.
```

Those SDT run as components on the PowerFlex storage layer.

## `powerflex` driver in LXD

The `powerflex` driver in LXD uses PowerFlex volumes for custom storage volumes, instances and snapshots.
For storage volumes with content type `filesystem` (containers and custom file-system volumes), the `powerflex` driver uses volumes with a file system on top (see [`block.filesystem`](storage-powerflex-vol-config)).
By default, LXD creates thin-provisioned PowerFlex volumes.

> LXD expects the PowerFlex protection domain and storage pool already to be setup.

Reviewer: ```suggestion
LXD expects the PowerFlex protection domain and storage pool already to be set up.
```

> Furthermore LXD assumes that it has full control over the storage pool.

Reviewer: ```suggestion
Furthermore, LXD assumes that it has full control over the storage pool.
```

Therefore, you should never maintain any volumes that are not owned by LXD in a PowerFlex storage pool, because LXD might delete them.

> This driver also behaves differently than other drivers in that it provides remote storage.

Reviewer: ```suggestion
This driver behaves differently than some of the other drivers in that it provides remote storage.
```

As a result and depending on the internal network, storage access might be a bit slower than for local storage.
On the other hand, using remote storage has big advantages in a cluster setup, because all cluster members have access to the same storage pools with the exact same contents, without the need to synchronize storage pools.

When creating a new storage pool using the `powerflex` driver, LXD tries to discover one of the SDT from the given storage pool.

> This SDT can be overwritten using [`powerflex.sdt`](storage-powerflex-pool-config).

Reviewer: ```suggestion
Alternatively, you can specify which SDT to use with [`powerflex.sdt`](storage-powerflex-pool-config).
```

Not sure I like "overwritten" here, because the driver probably doesn't even try to discover an SDT if the option is set?

Author: That is true, the wording is better this way.

LXD instructs the NVMe initiator to connect to all the other SDT when first connecting to the subsystem.

Due to the way copy-on-write works in PowerFlex, snapshots of any volume don't rely on its parent.

> As a result, volume snapshots are fully functional volumes themselves from whom additional snapshots can be taken.

Reviewer: ```suggestion
As a result, volume snapshots are fully functional volumes themselves, and it's possible to take additional snapshots from such volume snapshots.
```

> This tree of dependencies is called the PowerFlex vTree.

Reviewer: ```suggestion
This tree of dependencies is called the *PowerFlex vTree*.
```

Both volumes and their snapshots get added as standalone NVMe disks to the LXD host.

(storage-powerflex-volume-names)=
### Volume names

Due to a [limitation](storage-powerflex-limitations) in PowerFlex, volume names cannot exceed 31 characters.

> In order for LXD to identify the volumes type and snapshots, special identifiers are appended to the volumes name:

Reviewer: ```suggestion
In order for LXD to identify the volume's type and snapshots, special identifiers are appended to the volume's name:
```
Not sure if that is what you mean? I might also be "the volume types and snapshots" and "the volume names".

Author: It can actually be both. I would prefer to refer to the plural (your suggestion) since the remainder of the doc also looks to not address single volumes.

Type            | Identifier   | Example
:--             | :---         | :----------
Container       | `c_`         | `c(_project1)_c1`
Virtual machine | `v_`         | `v(_project2)_v1`
Image           | `i_`         | `i(_project3)_img`
Custom volume   | `u_`         | `u(_project4)_vol`

> When creating volumes in the `default` project its identifier gets omitted.

Reviewer: ```suggestion
When creating volumes in the `default` project, its identifier is omitted.
```

> This does not apply for custom storage volumes since their name may contain an underscore which could
> potentially lead to inconsistencies when running the `lxd recover` command.

Reviewer: ```suggestion
This does not apply for custom storage volumes, since their name may contain an underscore (which could
potentially lead to inconsistencies when running the `lxd recover` command).
```

Volumes for virtual machines and ISO files require another two characters.

> The VMs root volume contains the `.b` block suffix whilst an ISO file ends on `.i`.

Reviewer: Is that something the user needs to do? Or is this done automatically? If it's automatic, should it be in the table above?
Or does it result in an actual limit of 27 characters for choosing a VM/ISO name? If so, we should explicitly say so.

Author: That is done automatically by LXD. What about:

```suggestion
Volumes for virtual machines and ISO files contain another two extra characters.
The VMs root volume contains the `.b` block suffix whilst an ISO file ends on `.i`.
```

Reviewer: Maybe "Volumes for virtual machines and ISO files use two additional characters internally."?

> In order to have enough space for the snapshot identifier, at least two more characters need to be

Reviewer: ```suggestion
To have enough space for the snapshot identifier, at least two more characters must be
```

reserved. You can try keeping the snapshots name at a minimum by setting [`snapshots.pattern`](storage-powerflex-snapshot-pattern)

> to `%d`. This will result in the following volume names:

Reviewer: Again, we should specify what that actually means for the user (like "So if you plan on taking snapshots, don't use instance or volume names that exceed xx characters").

Author: I'll add another sentence that depicts that explicitly.

* Snapshot `snap0` for a virtual machine `v1`: `v_project1_v1@0`
* Snapshot `snap1` for a custom volume `vol`: `u_project1_vol@1`

(storage-powerflex-limitations)=
### Limitations

The `powerflex` driver has the following limitations:

Limit of snapshots in a single vTree

> : From any volume in PowerFlex only 126 snapshots can be taken due to an internal limitation in the PowerFlex vTree.

Reviewer: ```suggestion
: An internal limitation in the PowerFlex vTree does not allow to take more than 126 snapshots of any volume in PowerFlex.
```
Again, trying to word this in active voice. ;)

>   This limit also applies to any child of any of the parent volumes snapshots.

Reviewer: ```suggestion
  This limit also applies to any child of any of the parent volume's snapshots.
```
I assume there's only one parent volume? Otherwise it might be `parent volumes'`

Author: There can be a chain of snapshots like (the entire graph is the vTree): 

```
parent -> snap1 -> snap5 -> snap8
       -> snap2            
       -> snap3 -> snap6
                -> snap7
```

But effectively there is only one parent so `volume's` seems to be the right choice.

  A single vTree can only have 126 branches.

> Non optimized image storage

Reviewer: ```suggestion
Non-optimized image storage
```

: Due to the limit of 126 snapshots in the vTree, the PowerFlex driver doesn't come with support for optimized image storage.
  This would limit LXD to create only 126 instances from an image.

>   Instead, when launching a new instance the images contents get copied to the instances root volume.

Reviewer: ```suggestion
  Instead, when launching a new instance, the image's contents get copied to the instance's root volume.
```

Copying volumes
: PowerFlex does not support creating a copy of the volume so that it gets its own vTree.
  Therefore, LXD falls back to copying the volume on the local system.

>   This implicates an increased use of bandwidth due to the volumes contents being transferred over the network twice.

Reviewer: ```suggestion
  This implicates an increased use of bandwidth due to the volume's contents being transferred over the network twice.
```

Volume size constraints

> : In PowerFlex the size of a volume needs to be in multiples of 8 GiB.

Reviewer: ```suggestion
: In PowerFlex, the size of a volume must be in multiples of 8 GiB.
```

  This results in the smallest possible volume size of 8 GiB.
  However, if not specified otherwise, volumes are getting thin-provisioned by LXD.
  PowerFlex volumes can only be increased in size.

Volume name size limitation

> : In PowerFlex the name of a volume cannot exceed 31 characters.

Reviewer: ```suggestion
: In PowerFlex, the name of a volume cannot exceed 31 characters.
```

  LXD uses certain identifiers to indicate the purpose of a volume in the storage backend.
  See the section on [Volume names](storage-powerflex-volume-names) for more information.

Sharing custom volumes between instances
: The PowerFlex driver "simulates" volumes with content type `filesystem` by putting a file system on top of a PowerFlex volume.
  Therefore, custom storage volumes can only be assigned to a single instance at a time.

Sharing the PowerFlex storage pool between installations
: Sharing the same PowerFlex storage pool between multiple LXD installations is not supported.
  Due to the volume name limitations, no extra identifier can be added to distinguish between installations.

## Configuration options

The following configuration options are available for storage pools that use the `powerflex` driver and for storage volumes in these pools.

(storage-powerflex-pool-config)=
### Storage pool configuration

> Key                        | Type   | Default             | Description

Reviewer: Do we expect this PR to go in before or after https://github.com/canonical/lxd/pull/12645 ?
Either way, these options will need to be moved to the code eventually (which also gives us the option to add more information to some of the options).
I'll just do a quick review now since we'll need to fix it again later anyways.

Reviewer: The options should also be ordered alphabetically, but again, this will come for free with the other PR, so no need to fix now.

Author: Now that https://github.com/canonical/lxd/pull/12645 got merged I rebased from `main` and moved the config keys for volumes and pools into the code.

:--                        | :---   | :------             | :----------
`powerflex.gateway`        | string | -                   | Address of the PowerFlex Gateway

> `powerflex.gateway.verify` | bool   | `true`              | Whether or not to verify the PowerFlex Gateways certificate

Reviewer: ```suggestion
`powerflex.gateway.verify` | bool   | `true`              | Whether to verify the PowerFlex Gateway's certificate
```

`powerflex.user.name`      | string | `admin`             | User for PowerFlex Gateway authentication
`powerflex.user.password`  | string | -                   | Password for PowerFlex Gateway authentication

> `powerflex.pool`           | string | -                   | ID of the PowerFlex storage pool. In case you want to specify the storage pool via its name, also set `powerflex.domain`

Reviewer: ```suggestion
`powerflex.pool`           | string | -                   | ID of the PowerFlex storage pool (if you want to specify the storage pool via its name, also set `powerflex.domain`)
```

> `powerflex.domain`         | string | -                   | Name of the PowerFlex protection domain. Only required if `powerflex.pool` is specified using its name

Reviewer: ```suggestion
`powerflex.domain`         | string | -                   | Name of the PowerFlex protection domain (only required if `powerflex.pool` is specified using its name)
```

> `powerflex.mode`           | string | the discovered mode | Gets discovered automatically if the system provides the necessary kernel modules. Currently only `nvme` is supported.

Reviewer: ```suggestion
`powerflex.mode`           | string | the discovered mode | Gets discovered automatically if the system provides the necessary kernel modules; currently, only `nvme` is supported
```

`powerflex.sdt`            | string | one of the SDT      | PowerFlex NVMe/TCP SDT

> `powerflex.clone_copy`     | bool   | `false`             | Whether to not use PowerFlex snapshots when creating a snapshot of instances or custom volumes. See the [vTree limitations](storage-powerflex-limitations)

Reviewer: ```suggestion
`powerflex.clone_copy`     | bool   | `false`             | Whether to not use PowerFlex snapshots when creating a snapshot of instances or custom volumes (see the [vTree limitations](storage-powerflex-limitations))
```
Is it correct that if this is set to "true", we will NOT use PowerFlex snapshots?

Author: Yes, when you set `powerflex.clone_copy=true` on the storage pool the driver will create  cloned copies instead of snapshots when creating LXD volume snapshots.

Reviewer: The explanation feels a bit backwards - but let's fix this when we have more space than a field in a table. :)

Both volumes and their snapshots get added as standalone NVMe disks to the LXD host.

(storage-powerflex-volume-names)=
### Volume names

Due to a [limitation](storage-powerflex-limitations) in PowerFlex, volume names cannot exceed 31 characters.
In order for LXD to identify the volume types and snapshots, special identifiers are appended to the volume names:

Type            | Identifier   | Example
:--             | :---         | :----------
Container       | `c_`         | `c(_project1)_c1`
Virtual machine | `v_`         | `v(_project2)_v1`
Image           | `i_`         | `i(_project3)_img`
Custom volume   | `u_`         | `u(_project4)_vol`

When creating volumes in the `default` project, its identifier is omitted.
This does not apply for custom storage volumes, since their name may contain an underscore (which could
potentially lead to inconsistencies when running the `lxd recover` command).

Volumes for virtual machines and ISO files use two additional characters internally.
The VMs root volume contains the `.b` block suffix whilst an ISO file ends on `.i`.

To have enough space for the snapshot identifier, at least two more characters must be

> reserved. You can try keeping the snapshots name at a minimum by setting [`snapshots.pattern`](storage-powerflex-snapshot-pattern)

Reviewer: ```suggestion
reserved. You can try keeping the snapshot names at a minimum length by setting [`snapshots.pattern`](storage-powerflex-snapshot-pattern)
```
